### ISSUE TITLE
Feature request - factorize publisher to make it independent of lumbjerjack protocol

### ISSUE THREAD

**Issue by mheese** [driskell/log-courier#134]
Hi Jason,

I've been using your log-courier as a base for a log-courier --> Kafka implementation. You can track my status here https://github.com/mheese/log-courier and I have a first "working" version. I started the implementation as a transport plugin; that didn't work out though because of the tight integration of the lumberjack protocol (which I don't plan on using).

The implementation sucks though as it strips the current "src/lc-lib/publisher/publisher.go" implementation of all the lumberjack specific things, and also removes all acknowledgement of all sent payloads as I didn't really know how to deal with this. With a Kafka producer, you might want to run it in non-ack mode, which basically won't give you message delivery guarantees (accept from TCP-acks).

My request/idea is the following: if we could factorize the publisher code in the same manner that the transport factory is done, we could have a lumberjack publisher, and a raw publisher, and make transport plugins decide of what protocol they support. Then I could refactor my code and make a pull request. It is unfortunately impossible otherwise and it would be too bad if we couldn't get Kafka support into your mainline.

Besides from that: keep up the good work on log-courier! It is my favorite option for shipping file-based logs! And now there is a Kafka implementation on its way :) I'm using the native Go implementation for Kafka: https://github.com/Shopify/sarama . It works like a charm so far.

Cheers
Marcus

**Comment by driskell** [comment]
Hi Marcus,

I've seen your work already :) It was interesting to see someone taking it and trying something else.

My initial intention was Publisher for protocol and Transport for transport. Of course with Kafka this doesn't work as the wire protocol for Kafka is already defined.

I'm already about to start work to make TCP load balance like ZeroMQ. To do this Publisher is going to take on the work that ZeroMQ used to do with regards to queuing - in addition to keeping its retransmission and reconnection handling (which ZeroMQ duplicated.) Once this is done ZeroMQ transport becomes hugely irrelevant and I will likely remove it - the only real benefit would be slightly better performance with ZeroMQ 4 Curve encryption over TLS - but as it's not easily available on any of the platforms I target (CentOS/RedHat etc.) I don't see people using it. The fact ZeroMQ hides so much from me has become arduous too :-1: 

It's worth bearing in mind that I would like to, at all costs, keep the reliability guarantee (at-least-once) and offer encryption. If Kafka can do that that's great. It will just be worth knowing how Kafka handles sends - does it take care of connection handling? Does it queue? If so it may be the planned Publisher stuff needs moving too - in which case it may be that Kafka needs to be implemented as an alternative Publisher otherwise once protocol is shifted and once queuing/connection handling is shifted it's not really doing anything!

I'll do some of my own research at some point and see what the best way is. The Publisher/Transport areas are my next concentration though, along with implementing more tests in Go for those.

Jason

**Comment by driskell** [comment]
Just an update. I can see you split down the payload into individual events and push those to Kafka. Is there a specific reason you did this?

The protocol Log Courier uses over TLS/TCP/ZeroMQ is not lumberjack - it's a custom one which is designed as "messages" (so it's ZeroMQ compatible). For example there's a payload "message" containing a chunk of events, identified by an ID. Then there's an ACK "message" which contains a payload ID and how many events in that payload have been processed. Also a ping "message".

What would you think about pushing messages to Kafka rather than events? This would mean Kafka itself just acts as a transport mechanism - and that each entry is a "message" rather than a raw text "event".

(Regarding ping message - I'll abstract that so a future Kafka transport could ignore it - the only reason the ping message exists is to keep-alive the TCP connection when no data is moving, otherwise firewalls have a tendency to forget a quiet connection, causing timeouts when traffic resume again. I would assume Kafka system already takes care of this.)

**Comment by mheese** [comment]
Hi Jason,

after your 2 comments, I already have a far better understanding of the original code base. It's interesting to see that you're planning on dropping ZeroMQ - as this was the reason why I initially looked into log-courier. But I can definitely understand your reasons: the bad support for ZeroMQ among platforms (and I count myself as running RHEL and CentOS machines) is definitely a problem. One suggestion though, how one could solve it: Go is such an attractive language to operations, as it produces single-binaries. What I have done so far to support ZeroMQ 4 with encryption is to compile static libraries of zeromq and libsodium and linked log-courier against them, producing a single binary. That said, the build instructions are a bit of a pain but could be automated of course.

Why I looked at log-courier first was the support for a load-balanced transport mechanism - what zeromq provides. However, it turned out that in our company Kafka is a far better solution, as we don't only have log files, but also syslog and other sources, and Kafka provides a lot of operational decency when it comes to upgrading and scaling.

Coming to your questions:

1. Kafka supports acks of message deliveries in 3 ways: no acks (which still gives you a TCP ack), server-side ack of receiving messages, and server-side acks of all ISRs (in-sync-replicas) received and have written the messages. For my first quick & dirty implementation, I didn't care about acks, and depending on the scale of and load on the kafka cluster and the stream that you're producing, one actually might want to consider not using acks at all. This is why I think it has to be configurable at least. However, I definitely want to implement the other ack methods as well. The sarama library (golang Kafka library) provides you with a stream where all the acked messages are put.

2. Kafka takes care of connection handling. On purpose though (and that might be contrary to what you expect), Kafka clients have a connection timeout if they are idle for too long. Also Kafka in the current version (0.8.2.0) does not provide encryption, however, I have seen the patches for 0.9.x version for TLS support already, along with other security-related items (something I'm really looking forward to).

3. I purposely stripped out the - with what I thought was the lumberjack protocol - pendingPayload wrapper part. Without it, I can just plainly consume that queue with the current logstash-input-kafka plugin without the need of doing anything else on the input side. Basically it means, we don't need to write an own logstash input plugin, as there is already the Kafka plugin. I can see the necessity for the wrappwer with plain TCP transport though. However, if we can get around implementing a Kafka consumer in the logstash-input-log-courier plugin, that would save us work. Besides from that, another reason is that I'm consuming the queues not only from logstash, but also from other programs. And having another wrapping layer around the messages is a bit annoying. Additionally in your pendingPayload implementation you compress the messages (which in general is a good thing), but it is not configurable, which is another reason why I dropped that part, as it made the consuming part of the queue even more dependable on a particular consumer.

Hm, maybe making the publisher a factory and using 2 different publisher implementations is a way to go. But I'm open for suggestions. In the end I would like to see the kafka transport merged with log-courier somehow. So I'm open to any implementation suggestions :) Log-courier is just extremely good at harvesting files and tagging them upfront, and I haven't come across any other program which does it in such a nice way, this is why I chose to implement the Kafka producer with log-courier.

**Comment by driskell** [comment]
Thanks for the detailed feedback here! It's greatly appreciated.

Regarding load-balanced and ZeroMQ - one thing I realised is I can better manage ZeroMQ queues by having individual sockets per connection rather than the current one socket and let ZeroMQ handle load balancing. It means I can kill a socket to kill its queue, and know which connection failed and which payloads it had to handle resend. The logic for this essentially means a transport instance per server - and this can then carry onto TCP making that load-balance friendly. So ZeroMQ becomes purely performance benefit and (unlikely) routing benefit (you could throw in standard ZeroMQ proxies to route messages). So that's a bit more on the end-of-ZeroMQ saga!

Here's some responses to each of yours:

1. One of the main goals of LSF, which I carried over to Log Courier, has been to guarantee transmission. Whether or not I'd want to allow it to be turned off for a transport I'm not sure on yet, but to configure the level of ACK would definitely be acceptable. The default should be the regular server-side ACK - ISR can be an option. If there's good evidence that turning off ACK altogether increases performance then I will make it an option and note so in any documentation. But if there's no real performance increase then I'd not want to implement it as someone is bound to change it thinking it does impact it, losing something to gain nothing (Log Courier can send events constantly and asynchronously await acknowledgements so latency is already completely ruled out.) Implementation wise when turned off, we'd simply just process a "virtual ACK" after each send.

2. That's fine. I'll try to look more into the library to make sure I don't cause any issues with any future integration! Especially around timeouts. Encryption would be great but I've already decided that in Log Courier's case, it's not required - having been around the LSF repository for a long while there's lots of people who run this internally - myself included - and it's definitely far simpler when you don't need to mess with Certificate generation.

3. This goes to show how bad my code documentation has been :-) pendingPayload is intended to just provide tracking of a payload that is in transit. It allows Log Courier to keep track of payloads that have been sent (up to max_pending_payloads) but haven't had responses yet. It also processes ACK responses and strips out acknowledged events. The Generate() was just a helper to regenerate a payload message without any already acknowledged events - this is probably best moved to the TCP transport - in a Kafka case it can just read the event array and push each one on. Each ACK, as it returns, can be processed by the pendingPayload and it will knock out the acknowledged events one by one.

Also - you make a very good point about logstash-input-kafka. So I will rule out sending the courier messages via any future kafka transport and just have it support that. Anyone would be right in saying it's more intuitive that way ("kafka input can't receive kafka output from Log Courier?!")

So far I've started on the following:
* Publisher will simply handle load balancing across "endpoints" - waiting for an endpoint to say "ready" before passing it the next batch using a least-pending priority queue.
* An endpoint holds a transport instance and a set of pendingPayloads for that instance - it's purpose is to hold outstanding payloads and process ACKs
* Transports signal their endpoint when they are connected and ready for events, and pass ACKs back to its endpoint. The ready signal results in a single payload, so each time it receives a payload it can signal ready again to say "give me another" - or other logic might want it to wait for something.
* Transports can signal failure to the endpoint. This causes all pendingPayloads to be returned to Publisher for redistribution amongst other endpoints immediately. Once the endpoint recovers, it can simply signal Ready again to grab new events.

Things I'm still designing:
* Timeout handling - Thing is, it needs to be synchronised with event sending and ack receiving and the only place we have synchronisation here is in publisher - maybe transport needs some writing but I fear it may repeat code amongst transport implementations or potentially mean an extra layer and complexity.
* Keepalive handling - This definitely should be in the transport. It's closely related to timeout handling above though and I'm still wondering what the best way to get the synchronisation is.

Thanks for all the feedback and discussions - it's definitely helping in my refactoring of the transport layers :+1:

**Comment by mheese** [comment]
Hi Jason,

It has been really busy for me lately, but thank you so much for your long explanation and response! I have been actually reading your response and thinking about it as well. There is still much work that needs to be done on the Kafka transport side as well.

I'm thinking about using the feature that we can define custom fields on files for a Kafka purpose. E.g. imagine the following:
- I have a Kafka queue with 4 partitions, and I have 4 servers shipping logs
- with a Hash Partitioner I would be able to always ship logs from a particular server to a specific partition
- I could achieve that by adding a field to a log-courier message which has a "kafka_key" field for example

Also another example and purpose where I could use imagine it:
- you have differnt types of log files on servers normally: like postfix, apache, etc.
- depending on how you plan to process them in Kafka, you actually want to ship them to topics
- currently my implementation sets a fixed topic in the configuration file ... however, that doesn't provide you the flexibility of shipping to different topics with one and the same log-courier process
- if I could set the topic to use with an additional field, I could achieve that as well

Here is an example configuration json snippet:
```
...
        "files": [
                {
                        "paths": [ "/tmp/log-courier.log" ],
                        "fields": {
                                "type": "my_log_type",
                                "kafka": {
                                        "key": "partition1",
                                        "topic": "mykafkatopic"

...
```


What do you think about that idea?

Also, am I assuming correctly that I need to follow the publisher_transport_refactoring branch?

Thanks
Marcus

**Comment by driskell** [comment]
Hi Marcus,

Those are good thoughts. However, I do think an initial setup where it just has a single fixed topic will keep things simple and get things working, as it won't require too much change to the existing code - it will just be an addition.

Also, I do remember that most people wanting to use Kafka do so because they already have the system in place, so I would expect they'd just use a single topic for log shipping as they'd use separate topics for separate systems (I could be wrong though.)

Furthermore, the only reason I can see to use multiple topics would be to split logs across multiple Logstash clusters. I can only see this occur at extreme scale with a minority of people. Most people would be happy with a single cluster and single Logstash configuration. Usually all or nearly all logs will end up in the same output anyway (e.g. Elasticsearch). Although to be fair, I still think even at extreme scale, it would likely be better to still use a single cluster of Logstash, just make it bigger. It's a single configuration then and resources are fully utilised, where with multiple clusters it's easy for one to be under utilised and the other not.

Jason

**Comment by huksley** [comment]
Hello! I upvote to this feature. It would be nice to have GELF support in log-courier. For it to be possible, the protocol code must be separated from core.

**Comment by driskell** [comment]
@huksley GELF is a data format as opposed to a protocol. This might go deeper as Log Courier internally creates JSON events ("message" key is the line, and then there's various metadata such as host name, file path, file offset, and extra fields from config). Does GELF allow JSON?

Specifically this discussion is about protocol. What would the target be for some form of GELF transmission? Does it support at-least-once guarantees? I'm not in favour of implementing any transport which is not. But that doesn't stop you taking the code and implementing your own version :+1: I welcome that.

@mheese The branch is now the "2.0" branch which has other 2.0 features implemented. I'm now going to go back to 1.6 to do some maintenance on that for a 1.7 release. Then I'll get back to finishing off 2.0 which will have all the new changes. Protocol should be completely independent in 2.0 and support multiple endpoint simultaneously with failover and load balancing. Benchmarks show it runs same speed still which is good too!

On the topics bit - I still think single topic is best. At least to start with. I'm not so sure on whether multiple topics is the way to go. Logstash conditionals already cover all benefits as far as I can see.

**Comment by mheese** [comment]
Hi Jason,

Thank you very much for your input and I appreciate your effort developing towards that direction. I'll take some time over the weekend and merge your 2.0 branch with my code. Hopefully I can come up with something that makes it possible to do a proper pull request then later for version 2. I also had issues in the past days / weeks that sarama changed the kafka library API quite a lot, and it constantly was breaking builds. They have now issued releases, and I think I will continue to use the "stable" releases, so that it does not break things in the future so easily anymore.

I will also focus on implementing at-least-once guarantees. Actually, I want to give the user the option for the 3 types of delivery guarantees that Kafka offers: no acks, ack that it is written to at least one partition and acks that the message has been written to all ISRs (in-sync-replicas). For the first case (no acks), I am for now assuming that I am going to send "fake" acks back to your log-courier backend. 

Regarding the topics bit: multi topics is definitely something that I want to implement as I already have a use case for that. We are using Kafka here not only to feed Logstash (and ElasticSearch), and I have to handle quite an amount of different topics. So I would definitely benefit from that.

Regarding GELF: that sounds like a really interesting idea and would make log-courier even more versatile. I'm looking at log-courier not really as "ship log files to logstash" anymore, but more generally as a pretty good product for shipping log files to other products. Especially as it has that admin interface that can gives me good monitoring information, it qualifies as a good lightweight message shipping solution.

I would have other contributions, if you are interested in a pull request btw: RPM spec file, RHEL 6 SySV init script file and RHEL 7 systemd service file. I also saw that other people are looking for Windows builds that I could contribute (I'm creating Windows x64 executables including 64bit builds of ZeroMQ and libsodium).

I'm also thinking of writing an input for journald or syslog, but I will create another "Issue" for that to discuss that. Enjoy your weekend!

Cheers
Marcus

**Comment by driskell** [comment]
Thanks Marcus. Would still prefer the no-ack to be left out to prevent any conflict with my goal of having a minimum of at-least-once guarantees. I can't see it having any issue except performance which for 99% of users probably won't matter.

Regarding topics, I understand, it's just probably going to require changes elsewhere in Log Courier to implement it properly, so likely best left for next iteration. Only way I see is if you rely on a `fields` entry to set the topic but this in my eyes would be anti-intuitive and not the purpose of the `fields` config.

Regarding rpm spec, sysv and systemd - if you look in `contrib` folder I already have some :-) Feel free to compare and contrast though in another issue if you feel there can be improvements.

**Comment by driskell** [comment]
On inputs, yes another issue. I have been thinking of working on that but I think it will be post-2.0. It would be nice to feed from Windows Event Viewer for example on Windows - but to be fair Windows support has gone a little down hill lately as I need to do some proper testing (it's not one of my platforms so low on my list.)

**Comment by drocsid** [comment]
+1 for a kafka implementation! +1 for multiple topics!

**Comment by NikolaeVarius** [comment]
+1 for Kafka integration please.

Kafka is freaking amazing.

**Comment by groeneveld91** [comment]
+1 for Kafka integrations.

My 2 cents on the multiple topics front: I have two logstash clusters, stage and production, that will both consume each event exactly once, iff that event hasn't been tagged with "stage only". If it has been identified as a stage only event then it will 'only' go through the logstash-stage cluster. So one work around would be to 'drop' the stage only event when it gets into logstash-prod... but I don't like that. Was thinking of using different topics, and having logstash-stage subscribe to both the stage only topic and the shared topic, and logstash-prod only the shared topic. 

All in the vein of having a stage and prod ELK stack.

**Comment by driskell** [comment]
I'm not 100% with Kafka but if Courier pushed to one topic and stage and prod pulled, wouldn't the events be distributed rather than broadcast? Or does Kafka allow broadcasting one event to many consumers?

Exactly once is impossible it would be at the very best consumed "at least once"..

**Comment by kadaan** [comment]
A topic is broken up into partitions.  A producer sends data to partitions either directly via a hash of the key.  Consumers are part of a consumer group.  Topics can be consumed by multiple consumer groups.  Consumer groups independently track their offset within a topic.  Within a consumer group, a partition will only be consumed by one consumer.

**Comment by driskell** [comment]
Revisiting this, I actually think maybe the `fields` is indeed the best way to trigger changes in behaviour for a specific file's events inside the output. To get around the "unintuitiveness" you would just have a parameter called `topic` on the network config for Kafka and then actually use a field definition of some sort in it, like is used in Logstash, such as `${topic}` and this would be substituted from the message when it's submitted.

Will close this now though as lots has changed since this.
